### PR TITLE
Replace Font Awesome icons in Journalist Interface with PNGs

### LINKED ISSUE #5546: Consider replacing FontAwesome glyphs with PNGs
## Description

The Journalist Interface uses FontAwesome for various icons in the UI, which are rendered as Unicode placeholder glyphs is FontAwesome is not present, which is the case in Tor Browser's "Safest" security setting:

Screenshot from 2020-09-30 15-23-34

The current implementation is based on the reasoning that it's not necessary to use Tor Browser in "Safest" mode in the Journalist Interface, given defense in depth against unauthorized use (2FA login, authorization token). However, we do ask users to use "Safest" to visit the *Source Interface*, and journalists may make the assumption that they should do so for the *Journalist Interface* as well (we have little evidence one way or the other).

To avoid unexpected breakage, it seems reasonable to replace these glyphs with static PNG assets, rendered a high enough resolution to scale well on HiDPI displays.

### LINKED ISSUE #5576: JUI Web UI nits
## Just some Nina Nits
_Possibly to squeeze-into the forthcoming deletion flow work. Because I'm me, I had to._

- [ ] Capitalize the "F" in "Filter by codename"
- [ ] Replace the fontawesome icons with PNGs, for more graceful degradation. See tofu, below.

image

### PR DESCRIPTION
**Pull request by DrGFreeman**
## Status

Ready for review.

## Description of Changes

Fixes #5546
Fixes #5576

Changes proposed in this pull request:

**Replace Font Awesome icons with PNGs in Journalist Interface.**
    
Replace all Font Awesome icons in JI with PNGs so they are displayed correctly with Tor Browser in "Safest" setting (#5546 & #5576).  See screenshots in comment below.
    
**Related changes included:**
- Increase font size of source submission counts text in `/`
- Ensure consistent vertical spacing in users table in `/admin/`

**Other changes:**
- Capitalize "F" in "Filter by codename" placeholder text in `/` (#5576)
- Add "«" in "Back to admin interface" link text in `/admin/edit/<user_id>` for consistency with other admin pages.

## Testing

How should the reviewer test this PR?
Write out any special testing steps here.

**Journalist homepage `/`**

Login to Journalist Interface using Tor Browser with security settings set to "Safest".
- [ ] Verify "Filter by codename" placeholder text of input box has capital "F".
- [ ] Verify all icons display and are properly positioned.
- [ ] Hover on Delete button and verify white trash icon is visible on hover.
- [ ] Hover on source star and verify orange star is visible.
- [ ] Click star on a source and verify the star icon turns to orange.
- [ ] Verify nothing is abnormal with the page layout.

**Source submissions page `/col/<filesystem_id>`**

Login source interface in a new browser tab.
Submit one message.
Submit one file.

Refresh the JI browser tab.
Click on the new source's codename.
- [ ] Verify the chevron icon displays between the All Source link and the source codename.
- [ ] Verify all button icons display and are properly positioned.
- [ ] Hover on Delete Selected button and verify the white trash icon is visible.
- [ ] Verify nothing is abnormal with the page layout.
Download one source message.
Submit two replies.

Refresh Source Interface browser tab.
Delete one reply.

Refresh the JI browser tab.
- [ ] Verify all icons display left of the source codename (open envelope, closed envelope, reply (unread), check (read reply) and are properly positioned.
- [ ] Verify the rightmost icons display (files, messages, reply) and are properly positioned.
- [ ] Scroll to bottom and verify trash icon displays on DELETE SOURCE AND SUBMISSIONS button.
- [ ] Hover on DELETE SOURCE AND SUBMISSIONS button and verify that white trash icon is displayed.

**Admin page `/admin/`**

Click the Admin link.
- [ ] Verify the plus and pencil icons display in the ADD USER and INSTANCE CONFIG buttons.
- [ ] Verify the edit and trash icons display on each journalist row.
- [ ] Verify the vertical spacing between the rows is similar.
- [ ] Hover on the second journalist trash icon and verify the darker red trash icon displays.
- [ ] Verify nothing is abnormal with the page layout

**Add User page `/admin/add`**

Click the ADD USER button.
- [ ] Verify the plus icon displays in the ADD USER button.
- [ ] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.
Additional
**Edit User page `/admin/edit/<journalist_id>`**

Click the edit icon on one journalist row.
- [ ] Verify the "«" character displays at the begining of the *Back to admin interface* link.
- [ ] Verify the "sync" icon displays on the RESET PASSWORD, RESET MOBILE APP CREDENTIALS and RESET SECURITY KEY CREDENTIALS buttons.
- [ ] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.

**Instance Configuration page `/admin/config`**

Click the INSTANCE CONFIG button.
- [ ] Verify the bell icon displays on the SEND TEST OSSEC ALERT button.
- [ ] Verify the edit (pencil) icon displays on the UPDATE LOGO and UPDATE SUBMISSION PREFERENCES buttons.
- [ ] Verify nothing is abnormal with the page layout.

**Flag source for reply success page `/flag`**

Shutdown the server and temporarily change line 250 of `securedrop/source_app/main.py` from `if entropy_avail >= 2400:` to `if entropy_avail >= 2400 and False:`.
Restart the server, log in as a new source and submit a message.
View this new source via the Journalist Interface.
Click FLAG THIS SOURCE FOR REPLY.
- [ ] Verify the info circle icon is displays and is properly positioned.
- [ ] Verify nothing is abnormal with the page layout.

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

### If you made changes to documentation:

- [ ] Doc linting (`make docs-lint`) passed locally

### If you added or updated a code dependency:

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [ ] I would like someone else to do the diff review

### CODE REVIEW

**Comment by DrGFreeman** [comment]
Screenshots taken with Tor Browser at "Safest" setting below:

**JI homepage** (Updated)

image


**Source page**

image

**Admin interface**

image

**Add user page**

image

**Edit user page** (lower portion where icons are used)

image

**Config page**

image

**Flag source for reply success page**

image

**Comment by DrGFreeman** [comment]
## Notes

1. (Update: resolved in https://github.com/freedomofpress/securedrop/pull/5593/commits/161aebafe08c137eb3425aea51a3757fe61e6e7b) ~~Hover behavior on the source star in `/`, trash buttons and trash icon in `/admin/` is functional.  However, **I had to disable the tooltip on the trash icon in `/admin` to make the hover work**.  There is certainly a way to get it to work but this is beyond my knowledge of CSS at this point.~~
~~I suspect I would need to modify these [two selectors](https://github.com/DrGFreeman/securedrop/blob/ea36d78b56b0a2cd474a626f26183a0407a1edb0/securedrop/sass/modules/_button.sass#L118-L122) to also capture the images that are [inside the `<i>` element](https://github.com/DrGFreeman/securedrop/blob/ea36d78b56b0a2cd474a626f26183a0407a1edb0/securedrop/journalist_templates/admin.html#L43-L46) which is currently deactivated. I will read a bit on css selectors and try to make it work.~~

1. I have not cleaned up the repository of unused font-awesome files. I could include this in the changes if desired.

1. The `securedrop/static/i/` folder also contains many icons in PNG format I suspect are note used anymore (though some are used in flash messages). I could remove the unused icons if desired.

1. I put the new icons in `securedrop/static/icons` for now to avoid mixing up with the other icons in `securedrop/static/i` as per the previous note. I can move them inside `securedrop/static/i` if I clean the unused icons and ensure there is no filename collision.

1. (Update: addressed in https://github.com/freedomofpress/securedrop/pull/5593/commits/91b59e892c3bb593f785bcfe2a8fc5dc04c67c41) ~~The `securedrop/journalist_templates/flag.html` template uses the `fa-info-circle` icon. I have not replaced it yet as I could not figure how to access this page through the JI.~~

**Comment by DrGFreeman** [comment]
@eloquence, this is a new PR replacing #5591 where I unfortunately messed-up the git history trying to rebase the branch. This one has a cleaner commit history.

I also added the new icon files to the AppArmor profile as [you requested](https://github.com/freedomofpress/securedrop/pull/5591#issuecomment-712465560).

I added a test plan in the PR as well.

I will will keep on the tooltip mentioned in my note above and then switch the PR to "Ready for Review".

Also, please let me know your opinion about the other points in my notes above (i.e. cleanup of unused icons & fonts).

**Comment by eloquence** [comment]
> The securedrop/journalist_templates/flag.html template uses the fa-info-circle icon. I have not replaced it yet as I could not figure how to access this page through the JI.

Temporarily change line 250 `securedrop/source_app/main.py` from `if entropy_avail >= 2400:` to `if entropy_avail >= 2400 and False:`. Log in as a new source, and view this source via the Journalist Interface. Flag the source for reply, and you will see the confirmation message with the FA icon.

As you may have already noticed during development, until #5595 lands (and your PR is rebased onto it), hot reloading of the Journalist Interface does not work on current `develop`. That means that after making the change above, you have to restart the development environment for it to take effect.

**Comment by eloquence** [comment]
This looks overall great already, and the new icons look super slick (thanks @ninavizz for recommending these replacements) and work well at high zoom levels.

One issue I'm noticing in testing is that due to the lack of a `width` attribute on many of the PNG uses, you'll see layout reflows when the page is loaded or reloaded:

reload-issue

Note how everything jumps back and forth. You can prevent this by specifying both width and height consistently.

**Comment by DrGFreeman** [comment]
Thanks for the feedback @eloquence! I still have a lot to learn and this kind of feedback is very useful! I had put only one dimension on the icons since they did not scale to exact integer values on both dimensions. I'll add the missing dimensions (width mostly), I presume the sight distortion from the rounding will not be noticeable.

I restored the hover "tooltip" on the delete journalist icon in `/admin` in https://github.com/freedomofpress/securedrop/pull/5593/commits/161aebafe08c137eb3425aea51a3757fe61e6e7b.

I'll also replace the `fa-info-circle` icon, checking it with the method you provided.

Let me know if you also want me to include cleanup of unused icons and font awesome files as part of this PR.

Thanks again!

**Comment by DrGFreeman** [comment]
> As you may have already noticed during development, until #5595 lands (and your PR is rebased onto it), hot reloading of the Journalist Interface does not work on current `develop`. That means that after making the change above, you have to restart the development environment for it to take effect.

I had noticed. I got pretty good at pressing `Ctrl-C`, `Up Arrow`, `Enter` to restart the server then log back in the JI while working on this PR. I setup my KeePassXC to auto-fill the journalist credentials, including TOTP token to login quickly ;)

**Comment by DrGFreeman** [comment]
Last JI icon replaced: `fa-info-circle` in `/flag`. Screenshot added with others in first comment of PR.

Corresponding test steps added to PR as well.

@eloquence FYI, switching to "Ready for Review".

File: securedrop/journalist_templates/_source_row.html
Comment (rmol) [inline securedrop/journalist_templates/_source_row.html]: The `alt` attributes are a great addition, but we should translate them, so they're accessible in other languages. For example, `alt="{{ gettext("starred icon") }}"`.

File: securedrop/journalist_templates/_source_row.html
Comment (rmol) [inline securedrop/journalist_templates/_source_row.html]: Just writing this is causing me some self-loathing, but the messages icon is enough shorter than the files icon that it caught my eye. Compare it with a height of 15. To me they're better aligned, but the stretching does increase the weight of the lines in the messages icon. If no one else sees a need to change it I will happily agree. :man_shrugging: 

Height 13:

source-list-messages-icon-height-13

Height 15:

source-list-messages-icon-height-15

File: securedrop/journalist_templates/admin.html
Comment (rmol) [inline securedrop/journalist_templates/admin.html]: This would be an exception to my request to translate alt tags. Where the icon is a visual repetition of adjacent text, we should use `alt=""`, per [WCAG recommendations for decorative images](https://www.w3.org/WAI/tutorials/images/decorative/).

File: securedrop/journalist_templates/col.html
Comment (rmol) [inline securedrop/journalist_templates/col.html]: And here, I think I would use `alt=""` on the separator icon rather than have a screenreader say something like "link all sources greater than contemptible bodybuilding" (or whatever the source designation were, and as I don't have a screenreader set up at the moment, this is just a guess at what it might sound like).

Really, the breadcrumbs should be done in a more [semantic and accessible way](https://www.w3.org/TR/wai-aria-practices/#breadcrumb), but that's definitely out of scope for this PR. :slightly_smiling_face:

File: securedrop/journalist_templates/edit_account.html
Comment (rmol) [inline securedrop/journalist_templates/edit_account.html:7]: This is great for consistency, but it's made me realize that all the places we use this as a navigational symbol probably look pretty odd in languages that use it as a quotation mark. Our links must look like they have more to say. :laughing: 

That's _another_ issue to file.

File: securedrop/sass/modules/_cols.sass
Comment (rmol) [inline securedrop/sass/modules/_cols.sass:62]: Nice, relative units FTW.

**Review by rmol — CHANGES_REQUESTED** [review]
Very nice work. I have one suggestion about the height of the messages icon, and a request that you review the new `alt` tags -- there are quite a few that could probably be left empty, as they're equivalent to adjacent text, and those that should remain should also be marked for translation, so they're accessible in all our supported languages.

Functionally, this is great; I ran through the test plan and everything looks good.

**Journalist homepage `/`**

Login to Journalist Interface using Tor Browser with security settings set to "Safest".
- [x] Verify "Filter by codename" placeholder text of input box has
      capital "F". (NOTE: This is of course not visible at "Safest"; I
      verified it at "Standard".)
- [x] Verify all icons display and are properly positioned.
- [x] Hover on Delete button and verify white trash icon is visible on hover.
- [x] Hover on source star and verify orange star is visible.
- [x] Click star on a source and verify the star icon turns to orange.
- [x] Verify nothing is abnormal with the page layout.

**Source submissions page `/col/<filesystem_id>`**

Login source interface in a new browser tab.
Submit one message.
Submit one file.

Refresh the JI browser tab.
Click on the new source's codename.
- [x] Verify the chevron icon displays between the All Source link and the source codename.
- [x] Verify all button icons display and are properly positioned.
- [x] Hover on Delete Selected button and verify the white trash icon is visible.
- [x] Verify nothing is abnormal with the page layout.
Download one source message.
Submit two replies.

Refresh Source Interface browser tab.
Delete one reply.

Refresh the JI browser tab.
- [x] Verify all icons display left of the source codename (open envelope, closed envelope, reply (unread), check (read reply) and are properly positioned.
- [x] Verify the rightmost icons display (files, messages, reply) and are properly positioned.
- [x] Scroll to bottom and verify trash icon displays on DELETE SOURCE AND SUBMISSIONS button.
- [x] Hover on DELETE SOURCE AND SUBMISSIONS button and verify that white trash icon is displayed.

**Admin page `/admin/`**

Click the Admin link.
- [x] Verify the plus and pencil icons display in the ADD USER and INSTANCE CONFIG buttons.
- [x] Verify the edit and trash icons display on each journalist row.
- [x] Verify the vertical spacing between the rows is similar.
- [x] Hover on the second journalist trash icon and verify the darker red trash icon displays.
- [x] Verify nothing is abnormal with the page layout

**Add User page `/admin/add`**

Click the ADD USER button.
- [x] Verify the plus icon displays in the ADD USER button.
- [x] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.
Additional
**Edit User page `/admin/edit/<journalist_id>`**

Click the edit icon on one journalist row.
- [x] Verify the "«" character displays at the begining of the *Back
      to admin interface* link.
- [x] Verify the "sync" icon displays on the RESET PASSWORD, RESET MOBILE APP CREDENTIALS and RESET SECURITY KEY CREDENTIALS buttons.
- [x] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.

**Instance Configuration page `/admin/config`**

Click the INSTANCE CONFIG button.
- [x] Verify the bell icon displays on the SEND TEST OSSEC ALERT button.
- [x] Verify the edit (pencil) icon displays on the UPDATE LOGO and UPDATE SUBMISSION PREFERENCES buttons.
- [x] Verify nothing is abnormal with the page layout.

**Flag source for reply success page `/flag`**

Shutdown the server and temporarily change line 250 of `securedrop/source_app/main.py` from `if entropy_avail >= 2400:` to `if entropy_avail >= 2400 and False:`.
Restart the server, log in as a new source and submit a message.
View this new source via the Journalist Interface.
Click FLAG THIS SOURCE FOR REPLY.
- [x] Verify the info circle icon is displays and is properly positioned.
- [x] Verify nothing is abnormal with the page layout.

File: securedrop/journalist_templates/_source_row.html
Comment (eloquence) [reply securedrop/journalist_templates/_source_row.html]: My own bias is always to respect an icon's ratio if possible, but I'm fine either way. Flagging to @ninavizz for visibility.

Comment (DrGFreeman) [reply securedrop/journalist_templates/_source_row.html]: I had initially put the messages icon the same height as the files icon (16px, natural width of 17.9px) but found it appear much larger than the files icon (14 x 16px). I reduced it, probably too much.

Since this icon is naturally wider than high, I propose to put it at 16px wide and set its height at 14px (natural height would be 14.3px). This will give it the same area as the files icon except in "landscape" mode instead of "protrait" if I can use these terms. Below is a preview of the result:

image

File: securedrop/journalist_templates/edit_account.html
Comment (DrGFreeman) [reply securedrop/journalist_templates/edit_account.html:7]: I could remove the character on all admin pages. Let me know what you think.

File: securedrop/journalist_templates/_source_row.html
Comment (DrGFreeman) [reply securedrop/journalist_templates/_source_row.html]: `gettext` added for all non-blank `alt` attributes.

File: securedrop/journalist_templates/admin.html
Comment (DrGFreeman) [reply securedrop/journalist_templates/admin.html]: `alt=""` used where the alt text does not add information relative to what the adjacent text already provides.

File: securedrop/journalist_templates/col.html
Comment (DrGFreeman) [reply securedrop/journalist_templates/col.html]: `alt=""` used.

File: securedrop/journalist_templates/_source_row.html
Comment (DrGFreeman) [reply securedrop/journalist_templates/_source_row.html]: Above proposal implemented in 3d28682.

**Comment by DrGFreeman** [comment]
@rmol, I either set `alt=""` or used `gettext()` for the `alt` attributes as suggested. To help with the review, below is a visual map of where I set `alt=""` (red) or used `gettext()` (green):

image

image

image

image

image

image

image

Let me know if this is OK or if you'd like changes.

File: securedrop/journalist_templates/edit_account.html
Comment (rmol) [reply securedrop/journalist_templates/edit_account.html:7]: I've filed #5596 for that.

File: securedrop/journalist_templates/_source_row.html
Comment (rmol) [reply securedrop/journalist_templates/_source_row.html]: That looks better to me. @eloquence and @ninavizz?

**Comment by rmol** [comment]
> I either set `alt=""` or used `gettext()` for the `alt` attributes as suggested. To help with the review, below is a visual map of where I set `alt=""` (red) or used `gettext()` (green):
> [...]
> Let me know if this is OK or if you'd like changes.

OK? _OK_? :open_mouth: This is _fantastic_! Thank you so much for going to this trouble. I'm going to run through everything again with a screenreader, and this is going to make it so much easier to make sure I've tested all the changes.

**Review by rmol — APPROVED** [review]

**Comment by rmol** [comment]
I'm approving this, but will wait to hear what others think of the message icon sizing before merging. 

Having gone through these pages with a screenreader, the new alt text works pretty well. Right now you hear things like "download icon two unread", or "unstarred icon push button", but we have much bigger usability problems here for screenreader users, and if we address those I think we're going to want to change the alt text again to work better in a different context. The star/unstar buttons, for example, with the right alt text and ARIA attributes could be made to say something like "source stellar bronchoscope, star toggle button, not pressed", which gives a clearer idea of the button's current and potential state. 

Thank you again, @DrGFreeman, for the great work and making review so much easier. I really appreciate it.

**Comment by ninavizz** [comment]
@DrGFreeman These look phenomenal, THANK YOU!!

FWIW, some icons are taller than they are wide, and that's ok. The generalized footprint that icons occupy all taking-up the same volume of space, is generally more important. These look lovely. Merge away!

**Pull request merged**
